# Chapter 3. Linear Regression - Lab

- [3.6 Lab: Linear regression](#Lab:-Linear-regression)

## Lab: Linear Regression

In [1]:
import pandas as pd
import numpy as np
import statsmodels.api as sm
from patsy import dmatrices

In [2]:
boston_df = pd.read_csv('../data/boston.csv', index_col=0)

In [3]:
boston_df.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 506 entries, 1 to 506
Data columns (total 14 columns):
 #   Column   Non-Null Count  Dtype  
---  ------   --------------  -----  
 0   crim     506 non-null    float64
 1   zn       506 non-null    float64
 2   indus    506 non-null    float64
 3   chas     506 non-null    int64  
 4   nox      506 non-null    float64
 5   rm       506 non-null    float64
 6   age      506 non-null    float64
 7   dis      506 non-null    float64
 8   rad      506 non-null    int64  
 9   tax      506 non-null    int64  
 10  ptratio  506 non-null    float64
 11  black    506 non-null    float64
 12  lstat    506 non-null    float64
 13  medv     506 non-null    float64
dtypes: float64(11), int64(3)
memory usage: 59.3 KB


In [4]:
boston_df.describe().applymap('{:3.2f}'.format)

crim      zn   indus    chas     nox      rm     age     dis     rad  \
count  506.00  506.00  506.00  506.00  506.00  506.00  506.00  506.00  506.00   
mean     3.61   11.36   11.14    0.07    0.55    6.28   68.57    3.80    9.55   
std      8.60   23.32    6.86    0.25    0.12    0.70   28.15    2.11    8.71   
min      0.01    0.00    0.46    0.00    0.39    3.56    2.90    1.13    1.00   
25%      0.08    0.00    5.19    0.00    0.45    5.89   45.02    2.10    4.00   
50%      0.26    0.00    9.69    0.00    0.54    6.21   77.50    3.21    5.00   
75%      3.68   12.50   18.10    0.00    0.62    6.62   94.07    5.19   24.00   
max     88.98  100.00   27.74    1.00    0.87    8.78  100.00   12.13   24.00   

          tax ptratio   black   lstat    medv  
count  506.00  506.00  506.00  506.00  506.00  
mean   408.24   18.46  356.67   12.65   22.53  
std    168.54    2.16   91.29    7.14    9.20  
min    187.00   12.60    0.32    1.73    5.00  
25%    279.00   17.40  375.38    6.95   17.02  
50%    330.00   19.05  391.44   11.36   21.20  
75%    666.00   20.20  396.23   16.96   25.00  
max    711.00   22.00  396.90   37.97   50.00

In [5]:
boston_df = pd.get_dummies(boston_df, columns=['chas'])

In [6]:
boston_df.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 506 entries, 1 to 506
Data columns (total 15 columns):
 #   Column   Non-Null Count  Dtype  
---  ------   --------------  -----  
 0   crim     506 non-null    float64
 1   zn       506 non-null    float64
 2   indus    506 non-null    float64
 3   nox      506 non-null    float64
 4   rm       506 non-null    float64
 5   age      506 non-null    float64
 6   dis      506 non-null    float64
 7   rad      506 non-null    int64  
 8   tax      506 non-null    int64  
 9   ptratio  506 non-null    float64
 10  black    506 non-null    float64
 11  lstat    506 non-null    float64
 12  medv     506 non-null    float64
 13  chas_0   506 non-null    uint8  
 14  chas_1   506 non-null    uint8  
dtypes: float64(11), int64(2), uint8(2)
memory usage: 56.3 KB


In [7]:
predictors = ' + '.join(boston_df.drop('medv', axis=1).columns)
y, X = dmatrices('medv ~ '+str(predictors), return_type='dataframe', data=boston_df)
print(f'Number of predictors: {boston_df.shape[1]}')

Number of predictors: 15


## Data science questions and answers

Let's try to answer some questions concerning the _Boston_ data set.

1. Is there a relationship between the median house price and the predictors?

In [8]:
results = sm.OLS(y, X).fit()
print(results.summary())

                            OLS Regression Results                            
Dep. Variable:                   medv   R-squared:                       0.741
Model:                            OLS   Adj. R-squared:                  0.734
Method:                 Least Squares   F-statistic:                     108.1
Date:                Wed, 29 Dec 2021   Prob (F-statistic):          6.72e-135
Time:                        17:23:28   Log-Likelihood:                -1498.8
No. Observations:                 506   AIC:                             3026.
Df Residuals:                     492   BIC:                             3085.
Df Model:                          13                                         
Covariance Type:            nonrobust                                         
                 coef    std err          t      P>|t|      [0.025      0.975]
------------------------------------------------------------------------------
Intercept     25.2019      3.407      7.398      0.0

Due to the low $p$-value of the $F$-statistic, we can conclude that there is a clear relationship between the median house prices and the predictors.

2. How strong is the relationship?

The $R^2$ is 0.741, indicating that the model can explain over 74% of the variance in the data. The residual standard error using statsmodels is:

In [17]:
print(f'Residual standard error: {round(np.sqrt(results.mse_resid), 3)}')

Residual standard error: 4.745


The residual standard error can be alternatively derived from the calculation below.

In [18]:
print(f'Residual standard error: {round(np.sqrt(((y.values.ravel()-results.fittedvalues)**2).sum()/(y.shape[0]-14)), 3)}')

Residual standard error: 4.745


In [25]:
print(f'Average median house price: {round(np.mean(y.values), 3)}')

Average median house price: 22.533


In [33]:
print(f'Percentage error of the mean: {round(100*(np.sqrt(results.mse_resid)/np.mean(y.values)), 2)}%')

Percentage error of the mean: 21.06%
